arch/x86/kvm/vmx/vmx.c

```diff
@@ -3010,20 +3017,46 @@ void free_loaded_vmcs(struct loaded_vmcs *loaded_vmcs)
 {
 	if (!loaded_vmcs->vmcs)
 		return;
+#ifndef __PKVM_HYP__
 	loaded_vmcs_clear(loaded_vmcs);
 	free_vmcs(loaded_vmcs->vmcs);
+#else
+	/*
+	 * The pkvm hypervisor doesn't have smp call thus requires the vmcs is
+	 * cleared on the remote CPU when freeing it.
+	 */
+	BUG_ON(loaded_vmcs->cpu != -1);
+	pkvm_hyp_donate_host(__pkvm_pa(loaded_vmcs->vmcs), PAGE_SIZE, true);
+#endif
 	loaded_vmcs->vmcs = NULL;
 	if (loaded_vmcs->msr_bitmap)
+#ifndef __PKVM_HYP__
 		free_page((unsigned long)loaded_vmcs->msr_bitmap);
+#else
+		pkvm_hyp_donate_host(__pkvm_pa(loaded_vmcs->msr_bitmap), PAGE_SIZE, true);
+#endif
 	WARN_ON(loaded_vmcs->shadow_vmcs != NULL);
 }
 
 int alloc_loaded_vmcs(struct loaded_vmcs *loaded_vmcs)
 {
+#ifndef __PKVM_HYP__
 	loaded_vmcs->vmcs = alloc_vmcs(false);
+#endif
 	if (!loaded_vmcs->vmcs)
 		return -ENOMEM;
 
+#ifdef __PKVM_HYP__
+	if (pkvm_host_donate_hyp(__pkvm_pa(loaded_vmcs->vmcs), PAGE_SIZE)) {
+		loaded_vmcs->vmcs = NULL;
+		loaded_vmcs->msr_bitmap = NULL;
+		goto out_vmcs;
```

Comment:
Just a note: generally it would be good to return the error returned by pkvm_host_donate_hyp(), not just -ENOMEM for any error. But for now it would probably make our changes to KVM code here more intrusive, so for now it might be better to keep it as is.

Reply:
Yeah, agree.